File: src/main/java/com/streamsets/SessionsController.java
package com.streamsets;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;
import scala.Option;

@RestController
@RequestMapping("/sessions")
public class SessionsController {

    @Autowired
    private SessionManager sessionManager;

    @PostMapping("/create")
    private Mono<String> createSession(@RequestParam("expiresIn") long expiresIn) {
        return Mono.just( this.sessionManager.createSession( expiresIn ) );
    }

    @PostMapping("/validate")
    private Mono<SSOPrincipal> validate(@RequestParam("token") String token) {
        Option<SSOPrincipal> principal = this.sessionManager.validate( token );
        return principal.isDefined() ? Mono.just( principal.get()) : Mono.empty();
    }

    @PostMapping("/invalidate")
    private Mono<Boolean> invalidate(@RequestParam("token") String token) {
        return Mono.just( this.sessionManager.invalidate( token ) );
    }


}

File: src/main/scala/com/streamsets/spring/SessionsController.scala
package com.streamsets.spring

import com.streamsets.sch.{SSOPrincipal, SessionsManager}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.{PostMapping, RequestMapping, RequestParam, RestController}
import reactor.core.publisher.Mono

import java.util.Optional
import scala.jdk.OptionConverters.RichOption

@RestController
@RequestMapping(Array("/sessions"))
class SessionsController {
  @Autowired private val sessionManager: SessionsManager = null

  @PostMapping(Array("/create"))
  private def createSession(@RequestParam("expiresIn") expiresIn: Long): Mono[ String ] =
    Mono.just(this.sessionManager.createSessions(1, expiresIn).head)

  @PostMapping(Array("/validate"))
  private def validate(@RequestParam("token") token: String): Mono[ String ] = {
    Mono.fromFuture( this.sessionManager.validateAsync(token) )
      .map{
        case Some( principal ) => SSOPrincipal.serialize( principal )
        case None => "null"
      }
  }

  @PostMapping(Array("/invalidate"))
  private def invalidate(@RequestParam("token") token: String): Mono[Boolean] =
    Mono.just(this.sessionManager.invalidate(token))
}
